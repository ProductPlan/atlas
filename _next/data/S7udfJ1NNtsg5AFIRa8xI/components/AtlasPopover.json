{"pageProps":{"component":"AtlasPopover"},"__N_SSG":true}