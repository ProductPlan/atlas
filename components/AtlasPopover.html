<!DOCTYPE html><html><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width"/><title>ProductPlan Atlas</title><meta name="next-head-count" content="3"/><link rel="preconnect" href="https://fonts.googleapis.com"/><link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="true"/><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.1.2/css/all.min.css"/><link rel="shortcut icon" href="/atlas/favicon.ico"/><link rel="preconnect" href="https://fonts.gstatic.com" crossorigin /><link rel="preload" href="/atlas/_next/static/css/b3de8659b891ad42.css" as="style"/><link rel="stylesheet" href="/atlas/_next/static/css/b3de8659b891ad42.css" data-n-g=""/><link rel="preload" href="/atlas/_next/static/css/48309c2d926084e7.css" as="style"/><link rel="stylesheet" href="/atlas/_next/static/css/48309c2d926084e7.css" data-n-p=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/atlas/_next/static/chunks/polyfills-c67a75d1b6f99dc8.js"></script><script src="/atlas/_next/static/chunks/webpack-e0e074d834bfaea7.js" defer=""></script><script src="/atlas/_next/static/chunks/framework-c02f198d58d34563.js" defer=""></script><script src="/atlas/_next/static/chunks/main-88e61d9f66dc3fee.js" defer=""></script><script src="/atlas/_next/static/chunks/pages/_app-fa8438df928f083e.js" defer=""></script><script src="/atlas/_next/static/chunks/874-c64678d9ac232da4.js" defer=""></script><script src="/atlas/_next/static/chunks/917-19d7d6836e5ffd88.js" defer=""></script><script src="/atlas/_next/static/chunks/95-ae941e4def06298b.js" defer=""></script><script src="/atlas/_next/static/chunks/pages/components/%5Bcomponent%5D-01f7b89b64c2d201.js" defer=""></script><script src="/atlas/_next/static/YhycuUcO_ZoLxnTmgJVr7/_buildManifest.js" defer=""></script><script src="/atlas/_next/static/YhycuUcO_ZoLxnTmgJVr7/_ssgManifest.js" defer=""></script><style data-href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;700&display=swap">@font-face{font-family:'Inter';font-style:normal;font-weight:400;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcCO3FwrK3iLTeHuS_fvQtMwCp50KnMw2boKoduKmMEVuLyfMZs.woff) format('woff')}@font-face{font-family:'Inter';font-style:normal;font-weight:500;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcCO3FwrK3iLTeHuS_fvQtMwCp50KnMw2boKoduKmMEVuI6fMZs.woff) format('woff')}@font-face{font-family:'Inter';font-style:normal;font-weight:700;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcCO3FwrK3iLTeHuS_fvQtMwCp50KnMw2boKoduKmMEVuFuYMZs.woff) format('woff')}@font-face{font-family:'Inter';font-style:normal;font-weight:400;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa2JL7W0Q5n-wU.woff2) format('woff2');unicode-range:U+0460-052F,U+1C80-1C88,U+20B4,U+2DE0-2DFF,U+A640-A69F,U+FE2E-FE2F}@font-face{font-family:'Inter';font-style:normal;font-weight:400;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa0ZL7W0Q5n-wU.woff2) format('woff2');unicode-range:U+0301,U+0400-045F,U+0490-0491,U+04B0-04B1,U+2116}@font-face{font-family:'Inter';font-style:normal;font-weight:400;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa2ZL7W0Q5n-wU.woff2) format('woff2');unicode-range:U+1F00-1FFF}@font-face{font-family:'Inter';font-style:normal;font-weight:400;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa1pL7W0Q5n-wU.woff2) format('woff2');unicode-range:U+0370-03FF}@font-face{font-family:'Inter';font-style:normal;font-weight:400;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa2pL7W0Q5n-wU.woff2) format('woff2');unicode-range:U+0102-0103,U+0110-0111,U+0128-0129,U+0168-0169,U+01A0-01A1,U+01AF-01B0,U+1EA0-1EF9,U+20AB}@font-face{font-family:'Inter';font-style:normal;font-weight:400;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa25L7W0Q5n-wU.woff2) format('woff2');unicode-range:U+0100-024F,U+0259,U+1E00-1EFF,U+2020,U+20A0-20AB,U+20AD-20CF,U+2113,U+2C60-2C7F,U+A720-A7FF}@font-face{font-family:'Inter';font-style:normal;font-weight:400;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa1ZL7W0Q5nw.woff2) format('woff2');unicode-range:U+0000-00FF,U+0131,U+0152-0153,U+02BB-02BC,U+02C6,U+02DA,U+02DC,U+2000-206F,U+2074,U+20AC,U+2122,U+2191,U+2193,U+2212,U+2215,U+FEFF,U+FFFD}@font-face{font-family:'Inter';font-style:normal;font-weight:500;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa2JL7W0Q5n-wU.woff2) format('woff2');unicode-range:U+0460-052F,U+1C80-1C88,U+20B4,U+2DE0-2DFF,U+A640-A69F,U+FE2E-FE2F}@font-face{font-family:'Inter';font-style:normal;font-weight:500;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa0ZL7W0Q5n-wU.woff2) format('woff2');unicode-range:U+0301,U+0400-045F,U+0490-0491,U+04B0-04B1,U+2116}@font-face{font-family:'Inter';font-style:normal;font-weight:500;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa2ZL7W0Q5n-wU.woff2) format('woff2');unicode-range:U+1F00-1FFF}@font-face{font-family:'Inter';font-style:normal;font-weight:500;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa1pL7W0Q5n-wU.woff2) format('woff2');unicode-range:U+0370-03FF}@font-face{font-family:'Inter';font-style:normal;font-weight:500;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa2pL7W0Q5n-wU.woff2) format('woff2');unicode-range:U+0102-0103,U+0110-0111,U+0128-0129,U+0168-0169,U+01A0-01A1,U+01AF-01B0,U+1EA0-1EF9,U+20AB}@font-face{font-family:'Inter';font-style:normal;font-weight:500;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa25L7W0Q5n-wU.woff2) format('woff2');unicode-range:U+0100-024F,U+0259,U+1E00-1EFF,U+2020,U+20A0-20AB,U+20AD-20CF,U+2113,U+2C60-2C7F,U+A720-A7FF}@font-face{font-family:'Inter';font-style:normal;font-weight:500;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa1ZL7W0Q5nw.woff2) format('woff2');unicode-range:U+0000-00FF,U+0131,U+0152-0153,U+02BB-02BC,U+02C6,U+02DA,U+02DC,U+2000-206F,U+2074,U+20AC,U+2122,U+2191,U+2193,U+2212,U+2215,U+FEFF,U+FFFD}@font-face{font-family:'Inter';font-style:normal;font-weight:700;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa2JL7W0Q5n-wU.woff2) format('woff2');unicode-range:U+0460-052F,U+1C80-1C88,U+20B4,U+2DE0-2DFF,U+A640-A69F,U+FE2E-FE2F}@font-face{font-family:'Inter';font-style:normal;font-weight:700;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa0ZL7W0Q5n-wU.woff2) format('woff2');unicode-range:U+0301,U+0400-045F,U+0490-0491,U+04B0-04B1,U+2116}@font-face{font-family:'Inter';font-style:normal;font-weight:700;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa2ZL7W0Q5n-wU.woff2) format('woff2');unicode-range:U+1F00-1FFF}@font-face{font-family:'Inter';font-style:normal;font-weight:700;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa1pL7W0Q5n-wU.woff2) format('woff2');unicode-range:U+0370-03FF}@font-face{font-family:'Inter';font-style:normal;font-weight:700;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa2pL7W0Q5n-wU.woff2) format('woff2');unicode-range:U+0102-0103,U+0110-0111,U+0128-0129,U+0168-0169,U+01A0-01A1,U+01AF-01B0,U+1EA0-1EF9,U+20AB}@font-face{font-family:'Inter';font-style:normal;font-weight:700;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa25L7W0Q5n-wU.woff2) format('woff2');unicode-range:U+0100-024F,U+0259,U+1E00-1EFF,U+2020,U+20A0-20AB,U+20AD-20CF,U+2113,U+2C60-2C7F,U+A720-A7FF}@font-face{font-family:'Inter';font-style:normal;font-weight:700;font-display:swap;src:url(https://fonts.gstatic.com/s/inter/v12/UcC73FwrK3iLTeHuS_fvQtMwCp50KnMa1ZL7W0Q5nw.woff2) format('woff2');unicode-range:U+0000-00FF,U+0131,U+0152-0153,U+02BB-02BC,U+02C6,U+02DA,U+02DC,U+2000-206F,U+2074,U+20AC,U+2122,U+2191,U+2193,U+2212,U+2215,U+FEFF,U+FFFD}</style></head><body><div id="__next"><div><div class="Heading_Heading__EQADp"><a href="/atlas">üåê ProductPlan Atlas</a></div><div class="Documentation"><nav class="Navigation_Sidebar__ycYDm"><div class="Navigation_Sidebar__title__LoJn8">Atlas</div><ul class="Navigation_Sidebar__list__Nlfdq"><li class="Navigation_Sidebar__listItem__5Wet1"><a href="/atlas">Welcome</a></li><li class="Navigation_Sidebar__listItem__5Wet1"><a href="https://github.com/ProductPlan/atlas-mono" target="_blank" rel="noreferrer">GitHub <i class="fa-brands fa-github"></i></a></li><li class="Navigation_Sidebar__listItem__5Wet1"><a href="/atlas/tokens">Tokens</a></li></ul><div class="Navigation_Sidebar__title__LoJn8">Components</div><ul class="Navigation_Sidebar__list__Nlfdq"><li class="Navigation_Sidebar__listItem__5Wet1 Navigation_Sidebar__listItemActive__3t7we"><a href="/atlas/components/AtlasPopover">AtlasPopover</a></li><li class="Navigation_Sidebar__listItem__5Wet1"><a href="/atlas/components/BasicButton">BasicButton</a></li></ul></nav><div class="PageContent_Page__KQX_Q"><div class="PageContent_PageContent__Hwrl_"><h1 class="PageContent_PageHeading__H2sy3">AtlasPopover</h1><p class="PageContent_PageSubcopy__Ux2Ic">A popover which displays above or below the thing which opens it, pointing
back to that thing</p><div><h3>Example<!-- -->s</h3><div class="ComponentPage_Example__VsPS1"><header class="ComponentPage_ExampleHeader__jLmLE"><div><h4>Popover Above</h4></div><div><button class="IconButton"><i class="fa fa-code"></i></button></div></header><div style="text-align:center"><button>Show Popover Above</button></div></div><div class="ComponentPage_Example__VsPS1"><header class="ComponentPage_ExampleHeader__jLmLE"><div><h4>Popover Below</h4></div><div><button class="IconButton"><i class="fa fa-code"></i></button></div></header><div style="text-align:center"><button>Show Popover Below</button></div></div><div class="ComponentPage_Example__VsPS1"><header class="ComponentPage_ExampleHeader__jLmLE"><div><h4>Popover Left</h4></div><div><button class="IconButton"><i class="fa fa-code"></i></button></div></header><div style="text-align:center"><button>Show Popover Left</button></div></div><div class="ComponentPage_Example__VsPS1"><header class="ComponentPage_ExampleHeader__jLmLE"><div><h4>Popover Right</h4></div><div><button class="IconButton"><i class="fa fa-code"></i></button></div></header><div style="text-align:center"><button>Show Popover Right</button></div></div><div class="ComponentPage_Example__VsPS1"><header class="ComponentPage_ExampleHeader__jLmLE"><div><h4>Popover Simple</h4></div><div><button class="IconButton"><i class="fa fa-code"></i></button></div></header><div style="text-align:center"><button>Show Popover Simple</button></div></div><h3>Props</h3><table class="Table"><thead><tr><th class="Props_TableHeading__Kvchn">Name</th><th class="Props_TableHeading__Kvchn">Description</th><th class="Props_TableHeading__Kvchn">Type</th><th class="Props_TableHeading__Kvchn">Default</th><th class="Props_TableHeading__Kvchn">Required</th></tr></thead><tbody><tr><td class="Props_TableCell__zHnLs"><strong>placement</strong></td><td class="Props_TableCell__zHnLs">specify if it should be positioned above or below</td><td class="Props_TableCell__zHnLs">string</td><td class="Props_TableCell__zHnLs">&quot;top&quot;</td><td class="Props_TableCell__zHnLs Props_TableCellCenter__zg1dd"></td></tr><tr><td class="Props_TableCell__zHnLs"><strong>description</strong></td><td class="Props_TableCell__zHnLs">the primary text of the popover, truncated to 8 lines</td><td class="Props_TableCell__zHnLs">string</td><td class="Props_TableCell__zHnLs"></td><td class="Props_TableCell__zHnLs Props_TableCellCenter__zg1dd"><i class="fa fa-check-circle"></i></td></tr><tr><td class="Props_TableCell__zHnLs"><strong>title</strong></td><td class="Props_TableCell__zHnLs">text which displays in the header of the popover</td><td class="Props_TableCell__zHnLs">string</td><td class="Props_TableCell__zHnLs"></td><td class="Props_TableCell__zHnLs Props_TableCellCenter__zg1dd"></td></tr><tr><td class="Props_TableCell__zHnLs"><strong>onReadMoreClick</strong></td><td class="Props_TableCell__zHnLs">what to do if Read more is clicked. if not present, read more will not display</td><td class="Props_TableCell__zHnLs">func</td><td class="Props_TableCell__zHnLs"></td><td class="Props_TableCell__zHnLs Props_TableCellCenter__zg1dd"></td></tr><tr><td class="Props_TableCell__zHnLs"><strong>children</strong></td><td class="Props_TableCell__zHnLs">HTML/component(s) to display at the bottom of the popover, such as for metadata</td><td class="Props_TableCell__zHnLs">any</td><td class="Props_TableCell__zHnLs"></td><td class="Props_TableCell__zHnLs Props_TableCellCenter__zg1dd"></td></tr><tr><td class="Props_TableCell__zHnLs"><strong>simple</strong></td><td class="Props_TableCell__zHnLs">whether to display the default or simple popover</td><td class="Props_TableCell__zHnLs">bool</td><td class="Props_TableCell__zHnLs">false</td><td class="Props_TableCell__zHnLs Props_TableCellCenter__zg1dd"></td></tr><tr><td class="Props_TableCell__zHnLs"><strong>open</strong></td><td class="Props_TableCell__zHnLs">whether to show or hide the popover</td><td class="Props_TableCell__zHnLs">bool</td><td class="Props_TableCell__zHnLs">false</td><td class="Props_TableCell__zHnLs Props_TableCellCenter__zg1dd"></td></tr><tr><td class="Props_TableCell__zHnLs"><strong>referenceElement</strong></td><td class="Props_TableCell__zHnLs">the element the Popover should point to</td><td class="Props_TableCell__zHnLs">node</td><td class="Props_TableCell__zHnLs"></td><td class="Props_TableCell__zHnLs Props_TableCellCenter__zg1dd"></td></tr><tr><td class="Props_TableCell__zHnLs"><strong>offset</strong></td><td class="Props_TableCell__zHnLs">how far away the Popover should be from the element it points to</td><td class="Props_TableCell__zHnLs">number</td><td class="Props_TableCell__zHnLs">12</td><td class="Props_TableCell__zHnLs Props_TableCellCenter__zg1dd"></td></tr></tbody></table></div></div></div></div></div></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"component":"AtlasPopover"},"__N_SSG":true},"page":"/components/[component]","query":{"component":"AtlasPopover"},"buildId":"YhycuUcO_ZoLxnTmgJVr7","assetPrefix":"/atlas","isFallback":false,"gsp":true,"scriptLoader":[]}</script></body></html>