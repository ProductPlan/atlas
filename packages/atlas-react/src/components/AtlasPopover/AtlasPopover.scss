@import "~@productplan/atlas-styles/atlas.scss";

.AtlasPopover {
  display: none;
  border: 1px solid var(--a-color-light-gray-400);
  border-radius: 6px;
  font-size: var(--a-size-font-body-14);
  line-height: 20px;
  position: relative;
  width: 314px;
  background-color: white;
  filter: drop-shadow(0px 2px 12px rgba(40, 41, 41, 0.2));
  text-align: left;

  &--open {
    display: block;
  }

  &__header {
    background-color: var(--a-color-light-gray-100);
    border-radius: 6px 6px 0 0;
    font-weight: 500;
    padding: var(--a-spacers-mini) var(--a-spacers-small);
  }

  &__body {
    padding: var(--a-spacers-mini) var(--a-spacers-small);

    & > *:last-child {
      margin-bottom: 0;
    }
  }

  &__description {
    margin-bottom: var(--a-spacers-micro);
    display: -webkit-box;
    -webkit-line-clamp: 8;// TODO: do we ever want to customize this?
    -webkit-box-orient: vertical;
    overflow: hidden;
  }

  &__action {
    appearance: none;
    background: none;
    border: none;
    color: var(--a-color-blue-400);
    cursor: pointer;
    font-size: 1em;
    line-height: inherit;
    margin-bottom: var(--a-spacers-mini);
    padding: 0;
  }

  &__meta {
    margin: 0;
    padding: 0;

    & > :last-child {
      margin-bottom: 0;
    }

    & > * {
      margin-bottom: var(--a-spacers-micro);
    }
  }

  &::before {
    /* popover arrow */
    content: "";
    display: block;
    position: absolute;
  }
}

.AtlasPopover__arrow::after {
    content: "";
    display: block;
    position: absolute;
}


[data-popper-placement^="right"] > .AtlasPopover__arrow {
  left: -10px;
  border-right: 10px solid var(--a-color-light-gray-400);
  border-bottom: 10px solid var(--a-color-transparent-empty);
  border-top: 10px solid var(--a-color-transparent-empty);

  &::after {
    border-right: 9px solid var(--a-color-white-400);
    border-bottom: 9px solid var(--a-color-transparent-empty);
    border-top: 9px solid var(--a-color-transparent-empty);
    transform: translate(1px, -9px);
  }
}
[data-popper-placement^="left"] > .AtlasPopover__arrow {
  right: -10px;
  border-left: 10px solid var(--a-color-light-gray-400);
  border-bottom: 10px solid var(--a-color-transparent-empty);
  border-top: 10px solid var(--a-color-transparent-empty);

  &::after {
    border-left: 9px solid var(--a-color-white-400);
    border-bottom: 9px solid var(--a-color-transparent-empty);
    border-top: 9px solid var(--a-color-transparent-empty);
    transform: translate(-10px, -9px);

  }
}
[data-popper-placement^="top"] > .AtlasPopover__arrow {
  bottom: -8px;
  border-left: 10px solid var(--a-color-transparent-empty);
  border-right: 10px solid var(--a-color-transparent-empty);
  border-top: 10px solid var(--a-color-light-gray-400);
  height: 0;
  top: calc(100%);
  width: 0;

  &::after {
    border-left: 9px solid var(--a-color-transparent-empty);
    border-right: 9px solid var(--a-color-transparent-empty);
    border-top: 9px solid var(--a-color-white-400);
    transform: translate(-9px, -10px);
  }
}
[data-popper-placement^="bottom"] > .AtlasPopover__arrow {
  top: -11px;
  border-left: 10px solid var(--a-color-transparent-empty);
  border-right: 10px solid var(--a-color-transparent-empty);
  border-bottom: 10px solid var(--a-color-light-gray-400);
  bottom: calc(100%);
  height: 0;
  width: 0;
  top: -10px;

  &::after {
    border-left: 9px solid var(--a-color-transparent-empty);
    border-right: 9px solid var(--a-color-transparent-empty);
    border-bottom: 9px solid var(--a-color-light-gray-100);
    transform: translate(-9px, 1px);
  }
}
.AtlasPopover--simple > .AtlasPopover__arrow {
  display: none;
}
