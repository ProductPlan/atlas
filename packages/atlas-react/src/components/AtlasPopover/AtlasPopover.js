import React, { useState } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import "./AtlasPopover.scss";
import { usePopper } from "react-popper";

/** A popover which displays above or below the thing which opens it, pointing
 * back to that thing */
export default function AtlasPopover({
  description,
  title,
  onReadMoreClick,
  placement = "top",
  children,
  simple = false,
  open = false,
  referenceElement,
  offset = 12,
}) {
  const [popperElement, setPopperElement] = useState(null);
  const [arrowElement, setArrowElement] = useState(null);
  const { styles, attributes } = usePopper(referenceElement, popperElement, {
    placement: placement,
    modifiers: [
      {
        name: "offset",
        options: {
          offset: [0, offset],
        },
      },
      {
        name: "arrow",
        options: {
          element: arrowElement,
        },
      },
    ],
  });
  const classes = classNames("AtlasPopover", {
    "AtlasPopover--open": open,
    "AtlasPopover--simple": simple,
  });

  return (
    <>
      {open && (
        <div
          className={classes}
          ref={setPopperElement}
          style={styles.popper}
          {...attributes.popper}
        >
          {!simple && <header className="AtlasPopover__header">{title}</header>}
          <div className="AtlasPopover__body">
            <div className="AtlasPopover__description">{description}</div>
            {onReadMoreClick && (
              <button
                className="AtlasPopover__action"
                onClick={onReadMoreClick}
              >
                {"Read more"}
              </button>
            )}
            {children && !simple && (
              <div className="AtlasPopover__meta">{children}</div>
            )}
          </div>
          <div
            className="AtlasPopover__arrow"
            ref={setArrowElement}
            style={styles.arrow}
          />
        </div>
      )}
    </>
  );
}

AtlasPopover.propTypes = {
  /** specify if it should be positioned above or below */
  placement: PropTypes.string,
  /** the primary text of the popover, truncated to 8 lines */
  description: PropTypes.string.isRequired,
  /** text which displays in the header of the popover */
  title: PropTypes.string,
  /** what to do if Read more is clicked. if not present, read more will not display */
  onReadMoreClick: PropTypes.func,
  /** HTML/component(s) to display at the bottom of the popover, such as for metadata */
  children: PropTypes.any,
  /** whether to display the default or simple popover */
  simple: PropTypes.bool,
  /** whether to show or hide the popover */
  open: PropTypes.bool,
  /** the element the Popover should point to */
  referenceElement: PropTypes.node,
  /** how far away the Popover should be from the element it points to */
  offset: PropTypes.number,
};

/* Questions:
 * does a popover have a maximum height?
 */
