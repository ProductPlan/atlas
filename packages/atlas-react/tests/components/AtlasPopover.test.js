import React, { useState } from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import AtlasPopover from "@components/AtlasPopover";

describe("AtlasPopover", () => {
  it("renders", () => {
    function Wrapper() {
      const [refElement, setRefElement] = useState(null);
      return (
        <>
          <div style={{ height: "10px", width: "100px" }} ref={setRefElement} />
          <AtlasPopover
            title="Popover with defaults"
            description="Hello World"
            open={true}
            referenceElement={refElement}
          />
        </>
      );
    }
    const { container } = render(<Wrapper />);
    const content = screen.getByText("Hello World");
    expect(container).toMatchSnapshot();
    expect(content).toBeTruthy();
  });

  it("renders a read more action", () => {
    const handleClick = jest.fn();

    function Wrapper() {
      const [refElement, setRefElement] = useState(null);
      return (
        <>
          <div style={{ height: "10px", width: "100px" }} ref={setRefElement} />
          <AtlasPopover
            title="Popover with read more"
            description="Hello World"
            open={true}
            referenceElement={refElement}
            onReadMoreClick={handleClick}
          />
        </>
      );
    }
    const { container } = render(<Wrapper />);
    expect(container).toMatchSnapshot();

    const readMore = screen.getByText("Read more");
    expect(readMore).toBeTruthy();
    fireEvent.click(readMore);
    expect(handleClick).toHaveBeenCalled();
  });

  it("does not render children with type simple", () => {
    const handleClick = jest.fn();

    function Wrapper() {
      const [refElement, setRefElement] = useState(null);
      return (
        <>
          <div style={{ height: "10px", width: "100px" }} ref={setRefElement} />
          <AtlasPopover
            title="Popover simple"
            description="This is a description"
            open={true}
            referenceElement={refElement}
            onReadMoreClick={handleClick}
            simple={true}
          >
            <p>This text should not render</p>
          </AtlasPopover>
        </>
      );
    }
    const { container } = render(<Wrapper />);
    expect(container).toMatchSnapshot();

    expect(screen.getByText("Read more")).toBeTruthy();
    expect(screen.getByText("This is a description")).toBeTruthy();

    expect(screen.queryByText("This text should not render")).toBeNull();
    // give it a title and confirm there isnt a title
  });

  it("is not visible when open is false", () => {
    function Wrapper() {
      const [refElement, setRefElement] = useState(null);
      return (
        <>
          <div style={{ height: "10px", width: "100px" }} ref={setRefElement} />
          <AtlasPopover
            title="Popover with defaults"
            description="Hello World"
            open={false}
            referenceElement={refElement}
          />
        </>
      );
    }
    const { container } = render(<Wrapper />);
    expect(container).toMatchSnapshot();
    expect(screen.queryByText("Hello World")).toBeNull();
  });
});
