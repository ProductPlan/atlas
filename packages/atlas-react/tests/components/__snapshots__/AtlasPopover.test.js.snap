// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtlasPopover does not render children with type simple 1`] = `
<div>
  <div
    style="height: 10px; width: 100px;"
  />
  <div
    class="AtlasPopover AtlasPopover--open AtlasPopover--simple"
    style="position: absolute; left: 0px; top: 0px;"
  >
    <header
      class="AtlasPopover__header"
    >
      Popover simple
    </header>
    <div
      class="AtlasPopover__body"
    >
      <div
        class="AtlasPopover__description"
      >
        This is a description
      </div>
      <button
        class="AtlasPopover__action"
      >
        Read more
      </button>
      <div
        class="AtlasPopover__meta"
      >
        <p>
          This text should not render
        </p>
      </div>
    </div>
    <div
      class="AtlasPopover__arrow"
      style="position: absolute;"
    />
  </div>
</div>
`;

exports[`AtlasPopover is not visible when open is false 1`] = `
<div>
  <div
    style="height: 10px; width: 100px;"
  />
</div>
`;

exports[`AtlasPopover renders 1`] = `
<div>
  <div
    style="height: 10px; width: 100px;"
  />
  <div
    class="AtlasPopover AtlasPopover--open"
    style="position: absolute; left: 0px; top: 0px;"
  >
    <header
      class="AtlasPopover__header"
    >
      Popover with defaults
    </header>
    <div
      class="AtlasPopover__body"
    >
      <div
        class="AtlasPopover__description"
      >
        Hello World
      </div>
    </div>
    <div
      class="AtlasPopover__arrow"
      style="position: absolute;"
    />
  </div>
</div>
`;

exports[`AtlasPopover renders a read more action 1`] = `
<div>
  <div
    style="height: 10px; width: 100px;"
  />
  <div
    class="AtlasPopover AtlasPopover--open"
    style="position: absolute; left: 0px; top: 0px;"
  >
    <header
      class="AtlasPopover__header"
    >
      Popover with read more
    </header>
    <div
      class="AtlasPopover__body"
    >
      <div
        class="AtlasPopover__description"
      >
        Hello World
      </div>
      <button
        class="AtlasPopover__action"
      >
        Read more
      </button>
    </div>
    <div
      class="AtlasPopover__arrow"
      style="position: absolute;"
    />
  </div>
</div>
`;
